src/service/url-replacements-impl.js

```diff
@@ -571,8 +571,9 @@ export class GlobalVariableSource extends VariableSource {
 
     this.setAsync('AMP_STATE', key => {
       // This is safe since AMP_STATE is not an A4A whitelisted variable.
-      const {documentElement} = win.document;
-      return Services.bindForDocOrNull(documentElement).then(bind => {
+      const root = this.ampdoc.getRootNode();
+      const element = /** @type {!Element|!ShadowRoot} */(root.documentElement || root);
```

Comment:
The cast tells the typechecker that this will always be `{!Element|!ShadowRoot}`, but in the case what `documentElement` is falsy and `root` is used, `root` will be `ShadowRoot|Document`, and the call to `bindForDocOrNull` will fail.

I recommend using something like this
```
const element = dev().assertElement(root.documentElement || root.body || root);
```

Reply:
Thank you for the review. I'll adopt your suggested usage of `dev().assertElement()` and push a revision.

My thoughts about type cast were that `root.documentElement` will exist if `root` is a `Document`, resulting in an `Element`. If `root.documentElement` is falsy, then `root` should reliably be `ShadowRoot`, which is an acceptable param to `bindForDocOrNull()`: https://github.com/ampproject/amphtml/blob/3d8d00201aba7b34421b8c778c9893f3fa05eaa1/src/services.js#L170-L173

Reply:
Following-up on this, since your suggestion to use `.assertElement()` was based on the assumption that `bindForDocOrNull()` only accepts an `Element` type param, but `ShadowRoot` is supported, is any change actually needed here?

Reply:
Fair point, I missed that jsdoc. Thanks for taking the time to explain : ) Yes I believe no change will be needed then.